internal/storage/payload_writer.go

```diff
@@ -0,0 +1,509 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"sync"
+
+	"github.com/apache/arrow/go/v8/arrow"
+	"github.com/apache/arrow/go/v8/arrow/array"
+	"github.com/apache/arrow/go/v8/arrow/memory"
+	"github.com/apache/arrow/go/v8/parquet"
+	"github.com/apache/arrow/go/v8/parquet/compress"
+	"github.com/apache/arrow/go/v8/parquet/pqarrow"
+	"github.com/cockroachdb/errors"
+	"github.com/golang/protobuf/proto"
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/common"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+)
+
+var _ PayloadWriterInterface = (*NativePayloadWriter)(nil)
+
+type NativePayloadWriter struct {
+	dataType    schemapb.DataType
+	arrowType   arrow.DataType
+	builder     array.Builder
+	finished    bool
+	flushedRows int
+	output      *bytes.Buffer
+	releaseOnce sync.Once
+}
+
+func NewPurePayloadWriter(colType schemapb.DataType, dim ...int) (*NativePayloadWriter, error) {
+	var arrowType arrow.DataType
+	if typeutil.IsVectorType(colType) {
+		if len(dim) != 1 {
+			return nil, fmt.Errorf("incorrect input numbers")
+		}
+		arrowType = milvusDataTypeToArrowType(colType, dim[0])
+	} else {
+		arrowType = milvusDataTypeToArrowType(colType, 1)
+	}
+
+	builder := array.NewBuilder(memory.DefaultAllocator, arrowType)
+
+	return &NativePayloadWriter{
+		dataType:    colType,
+		arrowType:   arrowType,
+		builder:     builder,
+		finished:    false,
+		flushedRows: 0,
+		output:      bytes.NewBuffer(make([]byte, 0, 16<<20)),
```

Comment:
Not all field writer need buffer of this size, may cause unwant memory usage for during flush and compaction